pkg/localstore/sampler.go

```diff
@@ -84,6 +87,11 @@ func (db *DB) ReserveSample(
 
 	t := time.Now()
 
+	// protect the DB from any updates till we finish creating the sample
+	db.batchMu.Lock()
```

Comment:
On second thought, because the sampler can take several minutes, locking the function at the top may not be a good idea as it will delay everything else.

I propose making the addrChan a buffered channel with a large size (several million) and unlocking `batchMu` when the iterator finishes. 

By moving the mutex unlock to when the iterator finishes, we risk `db.get` returning nil entries as gc could be running in the background. Here I advise we introduce a second mutex called `gcMutex` that only the reserve sampler and the gc uses.

Reply:
If we go with a buffered channel of size 2^22 (the size of the reserve) it's a 134 mb overhead .